feature: Hook for checking cache status & reloadOnDowngrade

MartinCura opened the pull request:
Closes #17 

This PR makes CacheBuster into a Provider with a Context so that any distant child component can call `checkCacheStatus()` without needing the user to reload the page, effectively allowing them to maybe bust the cache whenever they want.

[Edit] I've also now added a new prop `reloadOnDowgrade`, which i think is quite useful.

Other small changes, which can be removed if you want:
* add `@ts-check` to top of file, so IDE tells on type erros without completely switching to TS
* remove arg from `window.location.reload()` as it was a type error (it has no arguments: https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-location-reload)
* wrap `checkCacheStatus` in a `useCallback()`
* updated `caniuse` in lockfile

**Please tell me if there's anything i can help you with to merge this. :)**

Example usage of new hook:

```jsx
// ...
import { useLocation } from 'react-router-dom';
import CacheBuster, { useCacheBuster } from 'react-cache-buster';

const isEnabled = process.NODE_ENV === 'production';

const AppUpdater = ({ children }: { children: ReactNode }) => {
    const VersionChecker = ({ children }: { children: ReactNode }) => {
        const [skippedFirstRender, setSkippedFirstRender] =
            useState<boolean>(false);
        const location = useLocation();
        const { checkCacheStatus } = useCacheBuster();

        // Every time the location changes, check for a new version
        // as this (for some apps) is an unobtrusive moment for this.
        useEffect(() => {
            if (!isEnabled) {
                return;
            }
            if (!skippedFirstRender) {
                setSkippedFirstRender(true);
                return;
            }
            checkCacheStatus();
        }, [location, checkCacheStatus]);

        return <>{children}</>;
    };

    return (
        <CacheBuster
            currentVersion={packageInfo.version}
            isEnabled={isEnabled}
            isVerboseMode={false}
            loadingComponent={<p style={{ margin: '20px' }}>Loading app...</p>}
            // really make sure browser doesn't cache meta.json
            metaFileDirectory={`/meta.json?t=${Date.now()}#`}
        >
            <VersionChecker>{children}</VersionChecker>
        </CacheBuster>
    );
};
```

MartinCura committed:
feat: hook with callback to checkCacheStatus

MartinCura committed:
chore: update caniuse

MartinCura committed:
nit: typo in README

MartinCura committed:
feat: reloadOnDowgrade prop to enable reloading for a downgrade

MartinCura commented:
Added new `reloadOnDowngrade` prop.

CagriAldemir commented:
Hi @MartinCura. Thank you for your development. I think it's a good idea too. I will review and publish the new version this week.

robmadden commented:
Hi there, thanks for creating this library.

I know it's only been one day, but is there any chance you could release this feature? I was hoping to use this library in one of my projects, thanks!

CagriAldemir commented on README.md:
```diff
@@ -60,7 +60,7 @@ const App = () => {
     >
 
-      //Your actual root component...
+      {/* // Your actual root component... */}
```
Can you revert this comment line to `// Your actual root component...` ?

CagriAldemir commented on README.md:
```diff
@@ -89,7 +89,8 @@ The process works as follows;
 | loadingComponent  | JSX      | :x:                | Component to be rendered during the new version control.                                                                                                                                                                                                                                                                                                  ...
 | metaFileDirectory | String   | :x:                | If public assets are hosted somewhere other than root on your server, you can pass the directory with this prop.                                                                                                                                                                                                                                          ...
-| onCacheClear      | Function | :x:                | This function is called before clearing the cache when a new version is found. If you pass this prop, cache clearing is not performed. Instead, the cache clearing function is sent as a parameter to this function and **you are expected to call this function**.<br/>Structure of the function: `oncacheClear: (refreshCacheAndReload: () => Promise<v ...
+| reloadOnDowngrade | Boolean  | :x:                | Whether to also bust the cache and reload if the version fetched from the server is lower.                                                                                                                                                                                                                                                                ...
```
Can you add `Default: **false**` to end of description?

CagriAldemir commented on src/CacheBuster.js:
```diff
@@ -76,7 +89,7 @@ function CacheBuster({
 
         log('The cache has been deleted.');
-        window.location.reload(true);
+        window.location.reload();
```
The "forceRefresh" parameter is still [supported by the Firefox browser](https://developer.mozilla.org/en-US/docs/Web/API/Location/reload#parameters). So you better add the parameter again.

MartinCura replied:
Missed that! Thanks for pointing it out.

CagriAldemir reviewed (changes requested):
I added some comments. Can you check them?

You also added a hook. Thank you for that. However, I would be very happy if you add the details about this hook to the README file and add a simple example of its use in the example project located in [this directory](https://github.com/CagriAldemir/react-cache-buster/tree/master/example).

MartinCura committed:
docs: show usage of new functionality in README and sample project

MartinCura committed:
fix: rollback removing `window.location.reload()`'s parameter

Still supported by Firefox.
https://developer.mozilla.org/en-US/docs/Web/API/Location/reload#parameters

MartinCura commented:
There you go, @CagriAldemir. Happy to change anything else as you wish, but also feel free to fix any nits yourself (i think you should be able to push to my branch) 😃 

We're already using this so very happy if we can merge it soon.

robmadden commented:
Thanks for your efforts!

robmadden commented:
Friendly ping: any update here?

MartinCura commented:
> I added some comments. Can you check them?
> 
> You also added a hook. Thank you for that. However, I would be very happy if you add the details about this hook to the README file and add a simple example of its use in the example project located in [this directory](https://github.com/CagriAldemir/react-cache-buster/tree/master/example).

One more gentle ping, sorry 😄 I made the small changes you asked for. Is there anything else i can do to get this merged? 😇 Thank you.

robmadden commented:
👋  any update here?

CagriAldemir commented:
@MartinCura @robmadden Hello there. I could not review due to my workload, but I have just published the new version of Cache Buster. Thanks for support.

robmadden commented:
@CagriAldemir no worries, thanks for following up!

MartinCura commented:
No worries, thank you so much ! Already updating this dep in our repo.
